substrate/runtime/contract/src/exec.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate. If not, see <http://www.gnu.org/licenses/>.
+
+use super::{CodeOf, ContractAddressFor, Module, Trait};
+use account_db::{AccountDb, OverlayAccountDb};
+use gas::GasMeter;
+use vm;
+
+use rstd::prelude::*;
+use runtime_primitives::traits::Zero;
+use runtime_support::StorageMap;
+use staking;
+use system;
+
+pub struct CreateReceipt<T: Trait> {
+	pub address: T::AccountId,
+}
+
+pub struct CallReceipt {
+	pub return_data: Vec<u8>,
+}
+
+pub struct ExecutionContext<'a, T: Trait + 'a> {
+	// typically should be dest
+	pub self_account: T::AccountId,
+	pub overlay: OverlayAccountDb<'a, T>,
+	pub depth: usize,
+}
+
+impl<'a, T: Trait> ExecutionContext<'a, T> {
+	/// Make a call to the specified address.
+	pub fn call(
+		&mut self,
+		caller: T::AccountId,
+		dest: T::AccountId,
+		value: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		_data: Vec<u8>,
+	) -> Result<CallReceipt, &'static str> {
+		let dest_code = <CodeOf<T>>::get(&dest);
+
+		// TODO: check the new depth
+
+		let call_base_fee = <Module<T>>::call_base_fee();
+		if gas_meter.charge(call_base_fee).is_out_of_gas() {
+			return Err("not enough gas to pay base call fee");
+		}
+
+		let (exec_result, change_set) = {
+			let mut overlay = OverlayAccountDb::new(&self.overlay);
+
+			if value > T::Balance::zero() {
+				transfer(
+					gas_meter,
+					false,
+					&self.self_account,
+					&dest,
+					value,
+					&mut overlay,
+				)?;
+			}
+
+			let mut nested = ExecutionContext {
+				overlay: overlay,
+				self_account: dest.clone(),
+				depth: self.depth + 1,
+			};
+			let exec_result = if !dest_code.is_empty() {
+				vm::execute(
+					&dest_code,
+					&mut CallContext {
+						ctx: &mut nested,
+						_caller: caller,
+					},
+					gas_meter,
+				).map_err(|_| "vm execute returned error while call")?
+			} else {
+				// that was a plain transfer
+				vm::ExecutionResult {
+					return_data: Vec::new(),
+				}
+			};
+
+			(exec_result, nested.overlay.into_change_set())
+		};
+
+		self.overlay.commit(change_set);
+
+		Ok(CallReceipt {
+			return_data: exec_result.return_data,
+		})
+	}
+
+	pub fn create(
+		&mut self,
+		caller: T::AccountId,
+		endowment: T::Balance,
+		gas_meter: &mut GasMeter<T>,
+		ctor: &[u8],
+		_data: &[u8],
```

Comment:
We're already using `&[u8]` here. So I think it may be better to change above L51 to use `&[u8]` as well to keep it consistent.

Reply:
Also possibly:

* L237 and L250.
* ` substrate/runtime/contract/src/lib.rs`: L120, L128 and L217.
* `substrate/runtime/contract/src/vm.rs`: L65.

Reply:
Changed!